Add byDim function

jmh530 opened the pull request:
Adds convenience functions byDim and apply. byDim provides iteration of a slice by dimension, while apply combines byDim and map to apply a function by dimension.

Functionality discussed in issue #58.

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+	
+    auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)(Slice!(kind, packs, Iterator) slice)
+    {
+        static assert (packs != [1], "1-dimensional slice not accepted");
```
this can be removed and generalised for vectors

jmh530 replied:
@9il I had only included that initially because transposed does not work properly on vectors, i.e.
```
auto slice = iota(3);
auto x = slice.transposed!1; //does not work
```
but with the adjustment for packs, then one can only pass dim=0 for a 1-dimensional vector and there is no error.

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+{
+    import mir.ndslice : Slice, SliceKind;
+	
+    auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)(Slice!(kind, packs, Iterator) slice)
```
please add `///`

jmh530 replied:
Do I also need to add the `///` before byDim?

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+    import mir.ndslice : Slice, SliceKind;
+	
+    auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)(Slice!(kind, packs, Iterator) slice)
+    {
```
```
if (dim < packs[0])
```

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+        {
+            return slice;
+        }
+        else
```
```d
...
else static if (dim + 1 < packs[0] || packs.length > 1)
{
            import mir.ndslice : canonical, transposed;
            return slice.canonical.transposed!dim;
} 
...
```
please cover this in unittest

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+	
+    auto slice = iota(3, 4);
+	
+    static immutable shape34 = [3, 4];
```
Please change to fixed size arrays (like `size_t[2] = shape34 = [3, 4];` ) for all unittests

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+@safe @nogc pure nothrow
+unittest
+{
+    import mir.ndslice : iota;
```
Please use detailed imports `mir.ndslice.topology: iota` for all cases

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
++/
+template byDim(size_t dim = 0)
+{
+    import mir.ndslice : Slice, SliceKind;
```
ditto, please use detailed imports `mir.ndslice.slice: ...`

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+{
+    import mir.ndslice : iota;
+	
+    auto slice = iota(3, 4);
```
please add comments to `iota` all iota declarations in documented unittests.
like this 
```d
//  0  1  2  3
//  4  5  6  7
//  8  9 10 11
``` 
This will be very helpful for all newcomers

9il commented on source/mir/ndslice/package.d:
```diff
@@ -116,12 +116,14 @@ $(TR $(TDNW $(SUBMODULE topology) $(BR)
      $(TD
-        $(SUBREF topology, as)
+        $(SUBREF topology, apply)
+	$(SUBREF topology, as)
```
tabs -> spaces

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+    auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)(Slice!(kind, packs, Iterator) slice)
+    {
+        static assert (packs != [1], "1-dimensional slice not accepted");
+		
```
Please remove trailing tabs and spaces

9il replied:
in all edited files

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+	$(LREF ipack),
+	$(LREF transposed).
++/
+template byDim(size_t dim = 0)
```
Please generalize it for multiple dimensions like it was done for `transposed`:
```d
template transposed(Dimensions...)
    if (Dimensions.length)
{
    static if (!allSatisfy!(isSize_t, Dimensions))
        alias transposed = .transposed!(staticMap!(toSize_t, Dimensions));
    else
  ....
```

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -66,6 +67,7 @@ $(T2 pack     , Returns slice of slices.)
 $(T2 ipack    , Returns slice of slices.)
 $(T2 unpack   , Merges all dimension packs.)
 $(T2 evertPack, Reverses dimension packs.)
+$(T2 byDim    , Convenience function to return a slice that can be iterated by dimension.)
```
Please also add a note what actually this function do: transposes dimensions on top and packs them.

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3069,411 @@ unittest
+    $(LREF map),
+    $(HTTP https://stat.ethz.ch/R-manual/R-devel/library/base/html/apply.html, apply)
++/
+template apply(alias f, size_t dim = 0)
```
I do not think we need this function. `slice.byDim!dim.map!f` is well enough. Also issue with `apply` name: `apply` is a verb in infinitive form, which is good for actual computations but not for lazy data type construction.

jmh530 replied:
Fair point on the naming, R's apply would really be more like `slice.byDim!dim.each!f`

9il commented:
ping @wilzbach

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,284 @@ unittest
+    import mir.ndslice.slice : Slice, SliceKind;
+
+    ///
+    auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)(Slice!(kind, packs, Iterator) slice)
```
Please add `@fastmath` flag for all inner functions

jmh530 commented:
@9il I think I have made all of the requested changes. 

Let me know if additional changes are required, or if you want me to rebase the three commits into one final one before merging (I already compressed a bunch into the two additional ones).

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,784 @@ unittest
+        alias byDim = .byDim!(staticMap!(toSize_t, Dimensions));
+    else
+    ///
+    auto byDim(SliceKind kind, size_t[] packs, Iterator)
```
`@fastmath`

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,784 @@ unittest
+}
+
+///
+@fastmath private auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)
```
no need `@fastmath` here because it is global for the module

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,784 @@ unittest
 }
+
+/// Transposed adjusted to ignore dim=0 and include universal
+@fastmath private template adjTransposed(Dimensions...)
```
no need @fastmath here because it is global for the module

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,784 @@ unittest
+        alias adjTransposed = .adjTransposed!(staticMap!(toSize_t, Dimensions));
+    else
+    ///
+    auto adjTransposed(SliceKind kind, size_t[] packs, Iterator)(Slice!(kind, packs, Iterator) slice)
```
@fastmath is required here because global fastmath does not affect internal template content

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,784 @@ unittest
+        {
+            return slice;
+        }
+        else
```
Please add specialization for `canonical` case. Canonical case is valid when the last (last of unpacked slice) dimension of contiguous slice is not moved upper. The code example for single dimension is below. Please generalize it for multiple dimensions.

```d
...
else static if (dim + 1 < packs[0] || packs.length > 1)
{
            import mir.ndslice : canonical, transposed;
            return slice.canonical.transposed!dim;
} 
...
```
please cover this in unittest (contiguous -> canonical) . Existing unittests should be enough, just extend them

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,784 @@ unittest
+       
+        mixin DimensionsCountCTError;
+            
+        static if (Dimensions.length == 1 && Dimensions[0] == 0)
```
`static if ([Dimensions] == [Iota!(Dimensions.length)])`
or 
`static if (Dimensions == Iota!(Dimensions.length))` but not sure if the last variant works

jmh530 commented:
@9il  I have fixed some of the issues...I definitely didn't understand your point on the canonical the first time you mentioned it. This second time I've at least added the code and adjusted it for multiple dimensions (though perhaps incorrectly). I still don't really understand the case, so I'm not really sure what I'm supposed to be testing for. So I didn't add any more tests and thus can't really be sure that it's currently right for multiple dimensions.

jmh530 commented:
@9il I've tried adding some more unittests (and some other small changes). Hopefully moving in the right direction.

As an aside, I have done some working with a function that formats an ndslice into a string (same format as the pretty versions I added here). At present I am making use of some features of the standard library and there is some use of the @gc. Does it have a place in mir-algorithm? Perhaps numir?

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,1452 @@ unittest
+       
+        mixin DimensionsCountCTError;
+
+        static if (packs == [1])
```
Please use the same multidimensional approach like in `adjTransposed`

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,1452 @@ unittest
+        import mir.ndslice.internal : DimensionsCountCTError, DimensionCTError;
+       
+        mixin DimensionsCountCTError;
+
```
just `return slice.adjTransposed!Dimensions.ipack!(Dimensions.length);`

9il replied:
please add a small unittest to cover it

jmh530 replied:
I had actually included a two-dimensional unittest on line 4412.

9il commented on source/mir/ndslice/topology.d:
```diff
@@ -3067,3 +3068,1452 @@ unittest
+}
+
+///
+auto byDim(SliceKind kind, size_t[] packs, Iterator)
```
I do not think we need this overload, because it is equal to `ipack!0`

jmh530 replied:
This was to allow writing the default x.byDim instead of x.byDim!0

9il commented:
> As an aside, I have done some working with a function that formats an ndslice into a string (same format as the pretty versions I added here). At present I am making use of some features of the standard library and there is some use of the @gc. Does it have a place in mir-algorithm? Perhaps numir?

Mir Algorithms need formatting, and it may be based on 100% generic @nogc code from Phobos. GC usage is allowed by user, but @nogc allocators should be supported by API.

For GC based code numic seems to be a good front-end library.

Thank you for byDim PR, good addition!

jmh530 commented:
@9il I think I could make progress on a @nogc version if issue #48 is resolved. The @gc part of my code is coming from std.array's join and appender.  

And I'm glad to help out. I feel like I'm getting a little better understanding of what you're doing here.

9il commented:
> @9il I think I could make progress on a @nogc version if issue #48 is resolved. The @gc part of my code is coming from std.array's join and appender.

If you have ability to fix #48 this would be awesome!

Phobos appender allocates GC struct on the heep. So it behaves as it is an object. Other approach is RefCounted appender.

Third approach is an appender with disabled `this(this)` constructor. This is fastest and simplest approach and it will work on 99.99% cases (it should be passed by `ref`). Plus it can be used to implement GC-allocated and ref-counted versions later.

With allocator based appender we have the following API issue:

```
struct Appender(Allocator) if (hasReallocate!Allocator ... ) { Allocator allocator; };
struct Appender(Allocator) if (hasReallocate!Allocator ... ) { Allocator* allocator; };
struct Appender(Allocator) if (hasReallocate!Allocator ... ) { /+Uses Allocator.instance+/ };
```

The last variant is valid only for some allocators like Mallocator and GCAllocator. The first may not be valid for some allocators (it is not specified yet, I just realize this issue, that allocators should have disabled `this(this)` too). The second variant is most universal.

This is draft:

```
struct Appender(T Allocator)
 if (hasReallocate!Allocator && ...)
{
    @disable this(this);
    private T[] _buffer;
    private size_t _length;

    static if (!hasInstance!Allocator)
    {
        private Allocator* _allocator;

        pragma(inline, true)
        private ref Allocator allocator()() { return *_allocator; }

        this(ref allocator)
        {
            _allocator = &allocator;
           // allocate buffer with say, 8 elements
        }
    }
    else
    {
        pragma(inline, true)
        private ref Allocator allocator()() { return Allocator.instance; }

        this(ref allocator)
        {
           // allocate buffer with say, 8 elements
        }
    }

    T[] data()()
    {
        return _buffer[0 .. _length];
    }

    void put()(T elem)
    {
        // ...
    }

    void put(Iterable)(Iterable iterable)
        if (is(ForeachType!Range : T))
    {
        // ...
    }
}
```

For now simple extend logic for `put` methods is OK (somthing like ). You can review/reuse Phobos code without any issue, we only will need to add proper credentials and list all authors.

Also, we need our `emplace` realization or we can use Phobos for now, or restrict types that can be used by Appender (builtin types,  classes, delegates, and structs without elaborate assign http://dlang.org/phobos/std_traits.html#hasElaborateAssign).

jmh530 commented:
@9il I'm not sure that I have the skills to get appender working properly (was just saying that if it did work, then I could do a @nogc format for slices). I'd say as a first step, I would need to re-write it to rely less on strings and more on ndslices. Then I would need to change use of std.format.format to std.format.sformat. I think only then would I check to see if I can get the appender working.

9il commented:
`sformat` throws :/ I like to have nothrow @nogc version of format at mir, it is quite large work to do from scratch, but we can use Phobos code, and merge our code back to phobos one day.

jmh530 commented:
Gotcha.
